File: frontend/src/lib/programPricing.js
/* The price a training program shows on the website — mirror of
 * backend/domains/program_pricing.py.
 *
 * The public endpoint already returns a formatted `pricing.display`, and every
 * customer-facing surface should print that rather than formatting a number
 * itself. This module exists for the two places that cannot read the endpoint:
 * the admin editor, which needs a live preview of what the website will say
 * before the program is saved, and any card rendering a program object that
 * predates the endpoint's projection.
 *
 * Nothing here derives a price. A program with no configured price reads as
 * "Contact us for pricing", which is a deliberate invitation rather than the
 * missing-data smell of the old "Ask for pricing".
 */

export const CONTACT = "contact";
export const PACKAGE = "package";
export const PER_UNIT = "per_unit";
export const FROM = "from";

export const CONTACT_COPY = "Contact us for pricing";

const UNITS = {
  session: "session",
  private_lesson: "private lesson",
  lesson: "lesson",
  week: "week",
  day: "day",
  night: "night",
  visit: "visit",
  month: "month",
};

/** $1,500 — not $1500 and not $1,500.00 when it's round. */
export function money(amount) {
  const value = Number(amount);
  if (!Number.isFinite(value)) return "";
  return Math.abs(value - Math.round(value)) < 0.005
    ? `$${Math.round(value).toLocaleString()}`
    : `$${value.toLocaleString(undefined, { minimumFractionDigits: 2, maximumFractionDigits: 2 })}`;
}

/**
 * What the public site will print for this program.
 *
 * Zero is not a price — it is the resting value of an untouched number input,
 * and it is how "$0" reached a customer-facing page before.
 */
export function publicPriceDisplay(program) {
  const mode = [CONTACT, PACKAGE, PER_UNIT, FROM].includes(program?.public_price_mode)
    ? program.public_price_mode
    : CONTACT;
  const raw = Number(program?.public_price_amount);
  const amount = Number.isFinite(raw) && raw > 0 ? raw : null;
  if (mode === CONTACT || amount === null) return CONTACT_COPY;
  if (mode === PACKAGE) return `${money(amount)} program`;
  if (mode === PER_UNIT) return `${money(amount)} / ${UNITS[program?.public_price_unit] || UNITS.session}`;
  return `From ${money(amount)}`;
}

/**
 * Prefer the server's own formatting when a program came from the public
 * endpoint; fall back to computing it. One string, whoever is asking.
 */
export function pricingLine(program) {
  const display = program?.pricing?.display;
  return typeof display === "string" && display ? display : publicPriceDisplay(program);
}

const UNIT_KEYS = Object.keys(UNITS);

/**
 * Why this pricing config can't be published, or null if it's fine.
 *
 * Mirrors backend/domains/program_pricing.assert_valid. Deliberately strict
 * about a priced mode with no amount: silently degrading to "Contact us" hides
 * the admin's own mistake, and they'd never know the website wasn't showing
 * the price they thought they'd set.
 */
export function publicPriceProblem(program) {
  const mode = program?.public_price_mode;
  if (!mode || mode === CONTACT) return null;
  if (![PACKAGE, PER_UNIT, FROM].includes(mode)) return `"${mode}" is not a pricing mode.`;
  const raw = program?.public_price_amount;
  if (raw === null || raw === undefined || raw === "") {
    return "Enter a price, or choose Contact us for pricing.";
  }
  const amount = Number(raw);
  if (!Number.isFinite(amount)) return "Enter the price as a number.";
  if (amount <= 0) return "Enter a price above zero, or choose Contact us for pricing.";
  if (mode === PER_UNIT && !UNIT_KEYS.includes(program?.public_price_unit)) {
    return "Choose what the rate is per (session, week, and so on).";
  }
  return null;
}

/**
 * The field changes that must accompany a pricing-mode switch.
 *
 * Switching to "Contact us" has to clear the amount, not just stop rendering
 * it: a stale number left on the record is one accidental mode-switch away
 * from being published as a price nobody agreed to. Likewise a unit only
 * means something for a per-unit rate.
 *
 * Returned as a patch rather than applied in the editor so the rule has one
 * definition and can be tested without mounting the studio.
 */
export function priceModePatch(mode, program) {
  const patch = { public_price_mode: mode };
  if (mode === CONTACT) {
    patch.public_price_amount = null;
    patch.public_price_unit = null;
    return patch;
  }
  if (mode !== PER_UNIT) {
    patch.public_price_unit = null;
  } else if (!UNIT_KEYS.includes(program?.public_price_unit)) {
    patch.public_price_unit = "session";
  }
  return patch;
}

File: backend/domains/program_pricing.py
"""What a training program costs, said only when somebody has actually said it.

The public Training page showed "Ask for pricing" on every program. That was
not a join failing — investigation found no program price has ever been
entered: five of the seven real programs have no `price` field at all and two
sit at 0.0, and the one program↔service link that exists
(`services.package_program_id`) is used by zero services.

The tempting fix is arithmetic: Level 1 is "5 sessions", a private lesson is
$90, therefore $450. That number would appear on the public website having
been invented here. A program package is not the sum of its parts — it can
carry a discount, a go-home session, boarding, follow-up support — so the
multiplication is a guess wearing a price tag.

So pricing is explicit or it is absent. Four modes, no inference:

    contact   — "Contact us for pricing" (the default, and not an error state)
    package   — one price for the whole thing:        "$1,200 program"
    per_unit  — a rate, with the unit spelled out:    "$90 / private lesson"
    from      — a floor, when the total genuinely varies: "From $450"

`contact` is the default on purpose. A program nobody has priced reads as a
deliberate invitation rather than as missing data, which is what "Ask for
pricing" felt like.
"""
from typing import Optional

CONTACT = "contact"
PACKAGE = "package"
PER_UNIT = "per_unit"
FROM = "from"

MODES = (CONTACT, PACKAGE, PER_UNIT, FROM)
DEFAULT_MODE = CONTACT

# Units a rate can be quoted in. Free text would let "/wk", "/week" and
# "/Week" all appear on one page.
UNITS = {
    "session": "session",
    "private_lesson": "private lesson",
    "lesson": "lesson",
    "week": "week",
    "day": "day",
    "night": "night",
    "visit": "visit",
    "month": "month",
}
DEFAULT_UNIT = "session"

CONTACT_COPY = "Contact us for pricing"


def _money(amount: float) -> str:
    """$1,500 — not $1500.0, and not $1,500.00 when it's a round number."""
    value = float(amount)
    if abs(value - round(value)) < 0.005:
        return f"${round(value):,}"
    return f"${value:,.2f}"


def is_valid_mode(mode: Optional[str]) -> bool:
    return mode in MODES


def normalize_unit(unit: Optional[str]) -> str:
    return unit if unit in UNITS else DEFAULT_UNIT


def resolve(program: dict) -> dict:
    """The public pricing line for one program.

    Returns ``{mode, display, amount, unit, has_price}``. ``display`` is the
    whole of what a public surface should print — no formatting decisions are
    left to the frontend, so the Training page, a program card and a share
    preview cannot drift apart.

    Anything incomplete falls back to Contact us rather than showing a half
    price. A mode of `package` with no amount is a misconfiguration, and the
    honest response to a misconfiguration is not $0.
    """
    mode = program.get("public_price_mode")
    if not is_valid_mode(mode):
        mode = DEFAULT_MODE

    amount = program.get("public_price_amount")
    try:
        amount = float(amount) if amount is not None else None
    except (TypeError, ValueError):
        amount = None
    # Zero is not a price. It is the default value of an untouched number
    # field, and it is how "$0" reached the public site in the first place.
    if amount is not None and amount <= 0:
        amount = None

    unit = normalize_unit(program.get("public_price_unit"))

    if mode == CONTACT or amount is None:
        return {"mode": CONTACT, "display": CONTACT_COPY, "amount": None,
                "unit": None, "has_price": False}
    if mode == PACKAGE:
        return {"mode": PACKAGE, "display": f"{_money(amount)} program", "amount": amount,
                "unit": None, "has_price": True}
    if mode == PER_UNIT:
        return {"mode": PER_UNIT, "display": f"{_money(amount)} / {UNITS[unit]}", "amount": amount,
                "unit": unit, "has_price": True}
    return {"mode": FROM, "display": f"From {_money(amount)}", "amount": amount,
            "unit": None, "has_price": True}


def assert_valid(program: dict) -> Optional[str]:
    """Error message if a submitted program's pricing is incoherent.

    Deliberately strict about a priced mode with no amount: silently showing
    "Contact us" for a program the admin believes they have priced is worse
    than telling them the field is empty.
    """
    mode = program.get("public_price_mode")
    if mode is None:
        return None
    if not is_valid_mode(mode):
        return f'"{mode}" is not a pricing mode. Choose one of: {", ".join(MODES)}.'
    if mode == CONTACT:
        return None
    amount = program.get("public_price_amount")
    try:
        amount = float(amount) if amount is not None else None
    except (TypeError, ValueError):
        return "Enter the price as a number."
    if amount is None or amount <= 0:
        return "Enter a price above zero, or choose Contact us for pricing."
    if mode == PER_UNIT and not is_valid_unit(program.get("public_price_unit")):
        return f'Choose a unit for the rate: {", ".join(sorted(UNITS))}.'
    return None


def is_valid_unit(unit: Optional[str]) -> bool:
    return unit in UNITS
